### ISSUE TITLE
RFC: Add `events` plugin to `@ngrx/signals`

### ISSUE THREAD

**Issue by markostanimirovic** [ngrx/platform#4580]
### Which @ngrx/* package(s) are relevant/related to the feature request?

signals

### Information

This RFC proposes adding the `events` plugin to the `@ngrx/signals` package to enable event-based state management with NgRx SignalStore.

## Key Principles

- Combines proven patterns from Flux, NgRx Store, and RxJS.
- Seamlessly integrates with existing SignalStore features.
- Extends Flux architecture with powerful customization options.
- Unifies local and global state management with a single approach.

## Prototype

The prototype of the `@ngrx/signals/events` plugin with a demo application is available at the following link: https://github.com/markostanimirovic/ngrx-signals-events-prototype

### Defining Events

Event creators are defined using the `eventGroup` function:

```ts
// users.events.ts

import { emptyProps, eventGroup, props } from '@ngrx/signals/events';

export const usersPageEvents = eventGroup({
  source: 'Users Page',
  events: {
    opened: emptyProps(),
    refreshed: emptyProps(),
  },
});

export const usersApiEvents = eventGroup({
  source: 'Users API',
  events: {
    usersLoadedSuccess: props<{ users: User[] }>(),
    usersLoadedFailure: props<{ error: string }>(),
  },
});
```

### Performing State Changes

The reducer is added to the SignalStore using the `withReducer` feature. Case reducers are defined using the `when` function:

```ts
// users.store.ts

import { when, withReducer } from '@ngrx/signals/events';

export const UsersStore = signalStore(
  { providedIn: 'root' },
  withEntities<User>(),
  withRequestStatus(),
  withReducer(
    when(usersPageEvents.opened, usersPageEvents.refreshed, setPending),
    when(usersApiEvents.usersLoadedSuccess, ({ users }) => [
      setAllEntities(users),
      setFulfilled(),
    ]),
    when(usersApiEvents.usersLoadedError, ({ error }) => setError(error)),
  ),
);
```

### Performing Side Effects

Side effects are added to the SignalStore using the `withEffects` feature:

```ts
// users.store.ts

import { Events, withEffects } from '@ngrx/signals/events';

export const UsersStore = signalStore(
  /* ... */
  withEffects(
    (
      _,
      events = inject(Events),
      usersService = inject(UsersService),
    ) => ({
      loadUsers$: events
        .on(usersPageEvents.opened, usersPageEvents.refreshed)
        .pipe(
          exhaustMap(() =>
            usersService.getAll().pipe(
              mapResponse({
                next: (users) => usersApiEvents.usersLoadedSuccess({ users }),
                error: (error: { message: string }) =>
                  usersApiEvents.usersLoadedError({ error: error.message }),
              }),
            ),
          ),
        ),
      logError$: events
        .on(usersApiEvents.usersLoadedError)
        .pipe(tap(({ error }) => console.log(error))),
    }),
  ),
);
```

Dispatched events can be listened to using the `Events` service.
If an effect returns a new event, it will be dispatched automatically.

### Reading State

State and computed signals are accessed via store instance:

```ts
// users.component.ts

@Component({
  selector: 'app-users',
  standalone: true,
  template: `
    <h1>Users</h1>

    @if (usersStore.isPending()) {
      <p>Loading...</p>
    }

    <ul>
      @for (user of usersStore.entities(); track user.id) {
        <li>{{ user.name }}</li>
      }
    </ul>
  `,
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class UsersComponent {
  readonly usersStore = inject(UsersStore);
}
```

### Dispatching Events

Events are dispatched using the `Dispatcher` service:

```ts
// users.component.ts

import { Dispatcher } from '@ngrx/signals/events';

@Component({
  /* ... */
  template: `
    <h1>Users</h1>

    <button (click)="onRefresh()">Refresh</button>

    <!-- ... -->
  `,
})
export class UsersComponent implements OnInit {
  readonly usersStore = inject(UsersStore);
  readonly dispatcher = inject(Dispatcher);

  ngOnInit() {
    this.dispatcher.dispatch(usersPageEvents.opened());
  }

  onRefresh(): void {
    this.dispatcher.dispatch(usersPageEvents.refreshed());
  }
}
```

It's also possible to define self-dispatching events using the `injectDispatch` function:

```ts
// users.component.ts

import { injectDispatch } from '@ngrx/signals/events';

@Component({
  /* ... */
  template: `
    <h1>Users</h1>

    <button (click)="dispatch.refreshed()">Refresh</button>

    <!-- ... -->
  `,
})
export class UsersComponent implements OnInit {
  readonly usersStore = inject(UsersStore);
  readonly dispatch = injectDispatch(usersPageEvents);

  ngOnInit() {
    this.dispatch.opened();
  }
}
```

### Scaling Up

The reducer can be moved to a separate file using the custom SignalStore feature:

```ts
// users.reducer.ts

export function withUsersReducer() {
  return signalStoreFeature(
    { state: type<EntityState<User> & RequestStatusState>() },
    withReducer(
      when(usersPageEvents.opened, usersPageEvents.refreshed, setPending),
      when(usersApiEvents.usersLoadedSuccess, ({ users }) => [
        setAllEntities(users),
        setFulfilled(),
      ]),
      when(usersApiEvents.usersLoadedError, ({ error }) => setError(error)),
    ),
  );
}
```

The same can be done for effects:

```ts
// users.effects.ts

export function withUsersEffects() {
  return signalStoreFeature(
    withEffects(
      (
        _,
        events = inject(Events),
        usersService = inject(UsersService),
      ) => ({
        loadUsers$: events
          .on(usersPageEvents.opened, usersPageEvents.refreshed)
          .pipe(
            exhaustMap(() =>
              usersService.getAll().pipe(
                mapResponse({
                  next: (users) => usersApiEvents.usersLoadedSuccess({ users }),
                  error: (error: { message: string }) =>
                    usersApiEvents.usersLoadedError({ error: error.message }),
                }),
              ),
            ),
          ),
        logError$: events
          .on(usersApiEvents.usersLoadedError)
          .pipe(tap(({ error }) => console.log(error))),
      }),
    ),
  );
}
```

The final SignalStore implementation will look like this:

```ts
// users.store.ts

export const UsersStore = signalStore(
  { providedIn: 'root' },
  withEntities<User>(),
  withRequestStatus(),
  withUsersReducer(),
  withUsersEffects(),
);
``` 

### I would be willing to submit a PR to fix this issue

- [X] Yes
- [ ] No

**Comment by rainerhahnekamp** [comment]
Well, I don't see ANY reason why this shouldn't land. 💯👍

**Comment by mauriziocescon** [comment]
Very nice! 👏🏿

I wonder: any specific reason why there isn't a `withEventGroup` or similar? I guess it would be nice to have everything together as well (`withEventGroup`, `withReducer`, `withEffects`).

**Comment by e-oz** [comment]
Is it something like global actions? I  mean, actions that any store can intercept.

**Comment by dmorosinotto** [comment]
Great work @markostanimirovic as @rainerhahnekamp stated just ship it 😉👍
Just one simple question about naming: with “self-dispatching” events using injectDispath you mean a “strong-typed” dispatcher!?🤔
 Because in the example I saw only “manual” callling dispatch.specificEvent(…) don’t understand the self-dispatching meaning or I miss something 🧐

**Comment by gabrielguerrero** [comment]
@markostanimirovic looks fantastic

**Comment by ducin** [comment]
Damn it, so I'll be the one "against" 😉

The main selling point behind NGRX-SS for me is its simplicity. The core (withState, withComputed, withMethods) is basically getting native angular stuff - but organizing it and making composable. Simple + composable + thin layer over native APIs -> win.

Nobody enforces to use a "plugin". But if majority of the community walked in redux style direction, I'd see it as a step backwards in terms of simplicity. Redux style includes significant tradeoffs and it does make sense to be used - but not in all scenarios. From my personal experience, I've seen redux far many more times abused than used properly. Hope that doesn't happen to NGRX signal store.

**Comment by CrazyJoker** [comment]
Looks great! What about returning multiple events in an effect, similar to using `mergeMap` or `concatMap` in @ngrx effects? Is this possible?

**Comment by mikezks** [comment]
Great work as always, @markostanimirovic.

This is exactly what we need and turns the Signal Store into a perfect state management solution, where devs may start lean and add the event-based API if needed later. 

> Very nice! 👏🏿
> 
> I wonder: any specific reason why there isn't a `withEventGroup` or similar? I guess it would be nice to have everything together (`withEventGroup`, `withReducer`, `withEffects`) whenever it makes sense.

Defining Events as with-Feature inside a Signal Store definition would not be a good idea as it leads to coupling of a single Signal Store instance to dispatch Events, which need to have a global nature in Redux-like patterns.

> Great work @markostanimirovic as @rainerhahnekamp stated just ship it 😉👍 Only one simple question about naming: with “self-dispatching” events using `injectDispath` you mean a “**strong-typed**” dispatcher!?🤔 Because in the example I saw only “manual” callling dispatch._specificEvent_(…) don’t understand the self-dispatching meaning, or I miss something?! 🧐

Self-dispatching means, that you do not need to call dispatch to pass in an Event, but just execute a method, pass in a payload if needed and the dispatching works automatically behind the covers.

**Comment by ValentinBossi** [comment]
@rainerhahnekamp and @markostanimirovic so so goooood🔥❤️
This was the missing part in signal store:
"**Unifies local and global state management with a single approach.**"

Having back a global event bus is so important🛠️

New people will use anyway signals in angular cause of angular docs and so will choose first the non redux way. so no worries about going in the wrong direction as those with redux needs would stay at ngrx store. this plugin just helps going in the right direction!

**Comment by davdev82** [comment]
I feel this change adds more complexity to Signal Store. For me the appeal of Signal Store, apart from reusability has been the simplicity of invoking methods. Adding Redux pattern to this adds more confusion for me. Why not use Ngrx then ? It can give all the benefits except reusability of features. And Ngrx also has selectSignal helpers to give signals out.

**Comment by marcindz88** [comment]
I think it's also a nice idea, if someone wants to inject a store that has both dispatch and get-data functionality, then its possible to define: 
```typescript
export const injectUserStore = () => ({ ...inject(UserStore), ...injectDispatch(userEvents) });
```
Maybe it would be nice to include some store-feature that merges action dispatcher into store or util / docs about it

**Comment by hudzenko** [comment]
Just discovered this feature at ng poland conf. This seems to be a game changer for our team for migration from global store to signal store. Waiting to be landed 🔥

**Comment by marcindz88** [comment]
Hi @markostanimirovic I think we should already consider a possibility to connect redux dev-tools.

I have a working hacky solution [here](https://github.com/marcindz88/ngrx-signals-events-prototype-with-dev-tools/commit/b3ff376b722593fad120714ea91c78d5483a22c4#diff-420b27aa750ba4d7d440fa484d1d47299c6da96afc88b6562c6ed3a66dbe0e6a), that could be used as an inspiration, but with the ability to change @ngrx internal packages, it would be possible to easily create a fully functional and quality solution.

**Comment by timdeschryver** [comment]
I like that events can be created outside of the store 🤩

**Comment by ValentinBossi** [comment]
> Looks great! What about returning multiple events in an effect, similar to using `mergeMap` or `concatMap` in @ngrx effects? Is this possible?

this is (when i understand you correctly) - see here https://ngrx.io/guide/eslint-plugin/rules/no-multiple-actions-in-effects - not recommended.

**Comment by BaronVonPerko** [comment]
I have mixed feelings about this. This is a great implementation of a redux approach to signal store, but I agree with @ducin, this takes away from the simplicity of what signal-store gives us.

Signals in general is a big step towards simplicity (computed is easier than dealing with rxjs pipes), and signal-store is already a great tool for organizing state with signals. 

With the reactivity we get from signals, I guess the question is, do we really need these events?

**Comment by ducin** [comment]
> do we really need these events?

💯. Or, slightly rephrased:

Do we need to put redux-alike impl. everywhere? (given that it is already implemented in so many places)

**Comment by e-oz** [comment]
Events are really needed sometimes, but they have a cost. Similar to `effect()`, they should be avoided, but they can sometimes be useful if used with care.
I recommend adding some guidelines to the documentation and JSDoc.

* Code that reacts to events **should only change internal state**, not something global or shared. For example, when an API service clears its cached requests because a global event declared the data is stale, that's a good use of events. When multiple services/components try to change URL params or trigger a redirect because of some event, that's a bad use of events.
* It becomes more difficult to trace the chain of calls. You cannot “click-through” methods in your IDE to see what causes a side effect. You need smart IDEs that can find event usages, and even then, you only see a list of usages, which might not clearly indicate what you're looking for.
* Events and listeners are fragile - when you refactor an event, you can't see direct consequences (as mentioned in the previous point). Over time and as the codebase grows, it becomes easier (or safer) to create a new event that “almost” duplicates an existing one. As you can see, this doesn't improve the overall code quality. Because of this, only global events should exist, and direct method calls should be preferred to internal events.
* Even within one store, if an event triggers mutations in different methods, it can create chaos, weird bugs, and unexpected behavior. Ideally, there should only be **one listener per event** in a store.

This is a list of reasons why I believe events should be avoided when possible and direct calls should be preferred.
However, there are examples of good event usage, and some cases can benefit noticeably from them.

**Comment by LcsGa** [comment]
I would definitely need it but I would never dispatch anything from the outside but instead, from within a feature that would trigger an effect on the parent like:
Both a pagination feature with infinite scroll and a category of product in a product store 
- if the category changes the page is reset to the first one + it loads x products
- if the page changes it loads x next products
- it loads the x first products on init

Would that be a valid use case or am I missing something (pretty new to stores tbh)?
> I tried with an effect in the onInit hook but it went into an infinite loop even though the only dependency there was the selectedCategory

**Comment by e-oz** [comment]
@LcsGa  I would use URL as the source of truth here. When the category is changed, `router.navigate(.., { queryParams: { category, page: 0}, queryParamsHandling: 'merge' });`.

Similar for pagination.

Then in the component that displays the items, I would use inputs, bound to query params, so they would change on navigation, and would load items using params.

But I could misunderstood your case.

**Comment by LcsGa** [comment]
Yep, that could be a way of handling this but it doesn't fit my use case:
- On top of the page I have a select for categories
- the rest of the same page is a grid with potentially infinite products, loading with an infinite scroll => when I reach the bottom of the page, I load the x next products of the selecte category

**Comment by e-oz** [comment]
@LcsGa component with the grid will not be destroyed/reloaded if you navigate to the same route. 
Having one source of truth (URL in this case) and letting users save/share their state of browsing using the URL is very valuable.

There are other ways, of course:
* Convert the store that keeps information about the selected category into a global store, and watch some observable/signal;
* Use events.

**Comment by LcsGa** [comment]
Yes I completely agree with you but unfortunately I can't do that due to my pagination, as it's made rn. If I had a paginator, I'd do that right away + it's maybe not the best example I gave but my purpose here is to stick with ngrx, to see if what I have in mind would be a good way to use events (I mean, if we didn't have the router option)

**Comment by yjaaidi** [comment]
🤯 this would make the SignalStore the most versatile state management ever.

I would emphasize that:
- it's an opt-in
- it should improve integration with devtools
- it enables the "shared event" pattern where one event can be handled by different stores/slices instead of creating interdependencies or manual indirection

what a team! 💪

**Comment by davdev82** [comment]
Really awesome and simple implementation. I would definitely like to give it a try. I like the last bit about scaling up where everything is so neat and concise.

**Comment by LcsGa** [comment]
Hey @markostanimirovic, any news about this package?

**Comment by Harpush** [comment]
> @LcsGa component with the grid will not be destroyed/reloaded if you navigate to the same route. Having one source of truth (URL in this case) and letting users save/share their state of browsing using the URL is very valuable.
> 
> There are other ways, of course:
> 
> * Convert the store that keeps information about the selected category into a global store (providedIn: root), inject it into the products grid component, and watch some observable/signal;
> * Use events.

I actually got a similar situation. If I have a pagination feature and I want to load from server on page change.

Option 1:
Similar to the new resource - invoke the load in an effect when the signals change. The downside is it can be hard to control when it runs.
Options 2:
The component calls update page. The feature update the page in state and emit an event. The main load rx method listens to the event (probably events) and execute the call.

What's best here?

**Comment by LcsGa** [comment]
I'd do the second one but until the events package becomes available in the signal-store package, it's not possible.
I had an issue with the first one : I had an effect when one specific value of my state changed but since the effect patches the state, it went in infinite loop (even if I use only one value from my state, it's still the whole state that changes and triggers the effect)

**Comment by rosostolato** [comment]
Any chance to make it work with not global stores? (with no providedIn root flag)

**Comment by mikezks** [comment]
> Any chance to make it work with not global stores? (with no providedIn root flag)

In the current PoC `Dispatcher` is an exported DI Token. So yes, if you provide it in another Injector node local Events can be used.

Nevertheless, I would first analyze if local dispatching is needed at all.

**Comment by rosostolato** [comment]
> > Any chance to make it work with not global stores? (with no providedIn root flag)
> 
> In the current PoC `Dispatcher` is an exported DI Token. So yes, if you provide it in another Injector node local Events can be used.
> 
> Nevertheless, I would first analyze if local dispatching is needed at all.

That would perfectly fit for stores with multiple instances. My current project would take great advantages of that.

**Comment by mikezks** [comment]
> That would perfectly fit for stores with multiple instances. My current project would take great advantages of that.

Technically, this would work. Especially if it comes to several instances it is questionable though if it really helpful to add the Event API here. You either need to create different Dispatcher Token or care about different Injector nodes. Both is not that easy to follow depending on the developer teams knowledge about Angular's Dependency Injection concept.

To summerize this, not everything that is technically possible makes sense from an architectural point of view. Most of the time the Event API will make sense for global State Management.

**Comment by rosostolato** [comment]
> @rosostolato, if you like to describe your use case in more detail, this could help to analyze whether additional features would be helpful to finalize the Event API.

Let me explain using my app as an example. In my app, there are multiple user workspaces, and they work kind of like Chrome tabs. Each workspace is like a separate tab, with its own data and its own store instance. So, when a user is working in a workspace, all the data for that workspace stays in its own store, completely separate from the others.

I know I could just use one global store for each entity to hold all the data and create selectors to filter out the entities for the current workspace. But trust me, that gets messy fast. It’s so much better to have individual stores where each one only holds the data for its specific workspace. This way, data from one workspace won’t interact with or accidentally affect another, and it also makes it much easier to create effects that only interact with the stores within the same workspace.

**Comment by audhit** [comment]
Hi any update on this matter? Really waiting for proper Redux Pattern for NGRX Signal Store. Currently doing it with "withRedux()" from ngrx-toolkit project, but unfortunately, it does not have any good separation of concern approach, and I am unable to separate my action, reducer and effects methods in their own corresponding files. So, would be great to know when NGRX team will release this feature approximately. Thanks :)

**Comment by rosostolato** [comment]
Any plans to resume this RFC?

**Comment by LcsGa** [comment]
Hi,

Following up on the discussion around scoping events to specific store instances (as raised by @rosostolato), I wanted to share an approach I've implemented for a similar requirement in my projects.

I often have scenarios with multiple component instances, each needing its own `SignalStore` instance. I required a way for external parts of the application (like other stores or services) to react to events emitted *specifically* by one of these instances.

To handle this, I created a custom, reusable event dispatcher utility. It's built using RxJS and leverages Angular's dependency injection for scoping.

Here's the core implementation:

```ts
type EventRecord = Record<string, unknown>; // { eventName: payloadType }

type ExcludeVoidPropertyKeys<T extends EventRecord> = {
  [K in keyof T]: T[K] extends void ? never : K;
}[keyof T];

type ExtractVoidPropertyKeys<T extends EventRecord> = {
  [K in keyof T]: T[K] extends void ? K : never;
}[keyof T];

/**
 * Creates a root injectable event dispatcher that can be used to dispatch and listen to events (described as the function type argument).
 *
 * (provided in root by default)
 */
export function createEventDispatcher<T extends EventRecord>(config?: { providedIn: 'root' | 'platform' | 'scoped' }) {
  type Event = keyof T;
  type DispatcherArg = { event: Event; payload?: T[Event] };

  return createInjectable(
    () => {
      const dispatcher$ = new Subject<DispatcherArg>();

      function dispatch<E extends ExtractVoidPropertyKeys<T>>(event: E): void;
      function dispatch<E extends ExcludeVoidPropertyKeys<T>>(event: E, payload: T[E]): void;
      function dispatch<E extends Event>(event: E, payload?: T[E]) {
        dispatcher$.next({ event, payload });
      }

      function on<E extends Event>(event: E) {
        return dispatcher$.asObservable().pipe(
          filter(({ event: e }) => e === event),
          map(({ payload }) => payload as T[E]),
        );
      }

      return {
        dispatch,
        on,
      };
    },
    { providedIn: config?.providedIn },
  );
}
```
> I used [`createInjectable`](https://ngxtension.netlify.app/utilities/injectors/create-injectable/) from ngxtension but simply using an `InjectionToken` would work too

Here's how I use it in practice:
1. Setup the event dispatcher alongside the store
```ts
// Use 'scoped' to ensure this dispatcher isn't root-provided,
// forcing us to provide it manually where needed (e.g., at the component level) for instance-scoping.
export const UserEvents = createEventDispatcher<{ loaded: User }>({ providedIn: 'scoped' });

export const UserStore = signalStore(
  withProps(() => ({ events: inject(UserEvents) })),
  withState<{ user: User | null }>({ user: null }),
  withMethods((store, gateway = inject(UserGateway)) => ({
    loadUser: rxMethod<...>(
      switchMap((...) => userService.getOne(...).pipe(
        tapResponse({
          next: (user) => {
            patchState(store, { user });
            store.events.dispatch('loaded', user);
          },
          error: (error) => ...
        })
      ))
    )
  }))
);
```
2. Provide and use the store and dispatcher at the component level
```ts
@Component({
  selector: 'app-user-profile',
  template: `...`,
  providers: [UserEvents, UserStore]
})
export class UserProfile {
  private readonly userStore = inject(UserStore);
  private readonly anotherStoreOrInjecatble = inject(...);

  constructor() {
    rxEffect(this.userStore.events.on('loaded'), (user) => this.anotherStoreOrInjecatble.doSomething(...));
  }
}
```
> I used [`rxEffect`](https://ngxtension.netlify.app/utilities/operators/rx-effect/) from ngxtension but a basic `.pipe(takeUntilDestroyed()).subscribe(...)` would work too

This pattern provides instance-specific event handling and satisfies my current needs quite well. It allows decoupling the store's internal actions from external reactions, scoped correctly via Angular's DI.
I could potentially build my own features like `withEffects` or `withReducers` on top of this if the need arises later and if the RFC is still ongoing (but I'd like to avoid going any further if possible).

@mikezks Does it give you any idea with that scoped events or do you see any issue(s) I missed?
Any update on this potential featur too?

**Comment by rohtashsethi-sophos** [comment]
@markostanimirovic Great work as always!

**Comment by rainerhahnekamp** [comment]
@LcsGa, events are usually global. Have you considered adding context information to an effect's payload? That could eliminate the necessity of scoped effects.

**Comment by rainerhahnekamp** [comment]
@LcsGa yeah, I would really manage that via the context. Otherwise, scoped effects could add unnecessary complexity to this extension.

**Comment by LcsGa** [comment]
Yeah maybe! And it is still possible to make it with a custom feature if I really feel the need for it 😉

**Comment by markostanimirovic** [comment]
PR for the `@ngrx/signals/events` plugin has been opened. It will be available in version 19.2.

**Comment by ValentinBossi** [comment]
❤️❤️❤️❤️❤️❤️❤️❤️❤️❤️

**Comment by markostanimirovic** [comment]
Thanks everyone for sharing your suggestions and specific use cases!

We'll investigate the concept of local events and come up with a solution in the next phase.

**Comment by ValentinBossi** [comment]
> PR for the `@ngrx/signals/events` plugin has been opened. It will be available in version 19.2.

❤️❤️❤️❤️❤️❤️❤️❤️❤️❤️

**Comment by markostanimirovic** [comment]
> Thanks everyone for sharing your suggestions and specific use cases!
> 
> We'll investigate the concept of local events and come up with a solution in the next phase.

Here is the proposal for local (scoped) events: https://github.com/ngrx/platform/issues/4776
